pgxn/neon/console_connector.c

```diff
@@ -0,0 +1,629 @@
+#include "postgres.h"
+#include "tcop/pquery.h"
+#include "tcop/utility.h"
+#include "access/xact.h"
+#include "utils/hsearch.h"
+#include "utils/memutils.h"
+#include "commands/defrem.h"
+#include "miscadmin.h"
+#include "utils/acl.h"
+#include "fmgr.h"
+#include "utils/guc.h"
+#include "port.h"
+#include <curl/curl.h>
+#include "utils/jsonb.h"
+
+static ProcessUtility_hook_type PreviousProcessUtilityHook = NULL;
+static char *ConsoleURL = NULL;
+
+typedef enum
+{
+    Op_Set,
+    Op_Delete,
+} OpType;
+
+typedef struct
+{
+    char name[NAMEDATALEN];
+    char owner[NAMEDATALEN];
+    char old_name[NAMEDATALEN];
+    OpType type;
+} DbEntry;
+
+typedef struct
+{
+    char name[NAMEDATALEN];
+    char old_name[NAMEDATALEN];
+    const char *password;
+    OpType type;
+} RoleEntry;
+
+typedef struct DdlHashTable
+{
+    struct DdlHashTable *prev_table;
+    HTAB *db_table;
+    HTAB *role_table;
+} DdlHashTable;
+
+static DdlHashTable RootTable;
+static DdlHashTable *CurrentDdlTable = &RootTable;
+
+static void PushKeyValue(JsonbParseState **state, int *estimated_len, char *key, char *value)
+{
+    JsonbValue k, v;
+    *estimated_len += strlen(key) + strlen(value) + 6; // Add 6 because of two pairs of quotes, a comma, and a colon
```

Comment:
can you calculate the length after building the whole JSON? because this 6 will look like magic if the space will be added somewhere

Reply:
It's just estimated length anyway, I could add just set it to 0 and correctness wouldn't be affected (it would just do a few extra memory allocations probably). So I wouldn't mind just setting `estimated_len` to 0 and just skip computing it.

Reply:
> So I wouldn't mind just setting estimated_len to 0 and just skip computing it.

+1. This isn't performance critical

Reply:
I got rid of this and set it to always be 0